feat(servicecatalog): Service Catalog is now in Developer Preview (#19204)
aws/aws-cdk: packages/@aws-cdk/aws-servicecatalog/README.md

- [Constraints](#constraints)
  - [Tag update constraint](#tag-update-constraint)
  - [Notify on stack events](#notify-on-stack-events)

>   - [CloudFormation template parameters constraint](#cloudformation-template-parameters-constraint)

Reviewer: Thoughts on `CloudFormation template constraint` instead?

Underlying SC terminology is that it is a Template Constraint and that Parameters are one of the examples a rule can apply to.

https://docs.aws.amazon.com/servicecatalog/latest/adminguide/reference-template_constraint_rules.html

Author: I think we've been moving towards more explicit naming.  I'm not actually sure if assertions can apply to anything but the parameters in a meaningful way?

## Portfolio

AWS Service Catalog portfolios allow administrators to organize, manage, and distribute cloud resources for their end users.
Using the CDK, a new portfolio can be created with the `Portfolio` construct:

```ts
new servicecatalog.Portfolio(this, 'Portfolio', {
  displayName: 'MyPortfolio',
  providerName: 'MyTeam',
});
```

> You can also specify optional metadata properties such as `description` and `acceptLanguage`

Reviewer 2: This is now called `messageLanguage`:

```suggestion
You can also specify optional metadata properties such as `description` and `messageLanguage`
```

to help better catalog and manage your portfolios.

```ts
new servicecatalog.Portfolio(this, 'Portfolio', {
  displayName: 'MyFirstPortfolio',
  providerName: 'SCAdmin',
  description: 'Portfolio for a project',
  messageLanguage: servicecatalog.MessageLanguage.EN,
});
```

Read more at [Creating and Managing Portfolios](https://docs.aws.amazon.com/servicecatalog/latest/adminguide/catalogs_portfolios.html).

To import an existing portfolio into your CDK application, use the `Portfolio.fromPortfolioArn()` factory method:

```ts
const portfolio = servicecatalog.Portfolio.fromPortfolioArn(this, 'ImportedPortfolio',
  'arn:aws:catalog:region:account-id:portfolio/port-abcdefghi');
```

### Granting access to a portfolio

> You can grant access to and manage the `IAM` users, groups, or roles that have access to the products within a portfolio.  Entities with granted access will be able to utilize the portfolios resources and products via the console or AWS CLI.

Reviewer 2: Multi-line please:

```suggestion
You can grant access to and manage the `IAM` users, groups, or roles that have access to the products within a portfolio. 
Entities with granted access will be able to utilize the portfolios resources and products via the console or AWS CLI.
```

Once resources are deployed end users will be able to access them via the console or service catalog CLI.

```ts fixture=basic-portfolio
import * as iam from '@aws-cdk/aws-iam';

const user = new iam.User(this, 'User');
portfolio.giveAccessToUser(user);

const role = new iam.Role(this, 'Role', {
  assumedBy: new iam.AccountRootPrincipal(),
});
portfolio.giveAccessToRole(role);

const group = new iam.Group(this, 'Group');
portfolio.giveAccessToGroup(group);
```

### Sharing a portfolio with another AWS account

> You can use account-to-account sharing to distribute a reference to your portfolio to other AWS accounts by passing the recipient account number. After the share is initiated, the recipient account can accept the share via CLI or console by importing the portfolio ID. Changes made to the shared portfolio will automatically propagate to recipients.

Reviewer 2: Multi-line please:

```suggestion
You can use account-to-account sharing to distribute a reference to your portfolio to other AWS accounts by passing the recipient account number.
After the share is initiated, the recipient account can accept the share via CLI or console by importing the portfolio ID.
Changes made to the shared portfolio will automatically propagate to recipients.
```

```ts fixture=basic-portfolio
portfolio.shareWithAccount('012345678901');
```

## Product

Products are version friendly infrastructure-as-code templates that admins create and add to portfolios for end users to provision and create AWS resources.
The CDK currently only supports adding products of type Cloudformation product. 
Using the CDK, a new Product can be created with the `CloudFormationProduct` construct.

> You can use `CloudFormationTemplate.fromUrl` to create a Product from a Cloudformation template directly from a URL that points to a CloudFormation template in S3, GitHub, or CodeCommit:

Author: ```suggestion
You can use `CloudFormationTemplate.fromUrl` to create a Product from a CloudFormation template directly from a URL that points to the template in S3, GitHub, or CodeCommit:
```

### Creating a product from a local asset

> A `CloudFormationProduct` can also be created by using a Cloudformation template held locally on disk using Assets.

Author: ```suggestion
A `CloudFormationProduct` can also be created by using a CloudFormation template held locally on disk using Assets.
```

### Creating a product from a stack

You can create a Service Catalog `CloudFormationProduct` entirely defined with CDK code using a service catalog `ProductStack`.
A separate child stack for your product is created and you can add resources like you would for any other CDK stack,
such as an S3 Bucket, IAM roles, and EC2 instances. This stack is passed in as a product version to your

> product.  This will not create a separate Cloudformation stack during deployment. 

Reviewer: `CloudFormation` just for consistency

> product.  This will not create a separate Cloudformation stack during deployment. 

Author: ```suggestion
product.  This will not create a separate CloudFormation stack during deployment. 
```

### Adding a product to a portfolio

> You add products to a portfolio to organize and distrbute your catalog at scale.  Addinga product to a portfolio creates an association, 

Author: ```suggestion
Adding a
```

> You add products to a portfolio to organize and distrbute your catalog at scale.  Addinga product to a portfolio creates an association, 

Reviewer: a few typos here

> You add products to a portfolio to organize and distrbute your catalog at scale.  Addinga product to a portfolio creates an association, 

Reviewer: ```suggestion
You add products to a portfolio to organize and distribute your catalog at scale.  Adding a product to a portfolio creates an association, 
```

and the product will become visible within the portfolio side in both the Service Catalog console and AWS CLI.
You can add a product to multiple portfolios depending on your organizational structure and how you would like to group access to products.

```ts fixture=portfolio-product
portfolio.addProduct(product);
```

## Tag Options

TagOptions allow administrators to easily manage tags on provisioned products by providing a template for a selection of tags that end users choose from.
TagOptions are created by specifying a tag key with a set of allowed values and can be associated with both portfolios and products.
When launching a product, both the TagOptions associated with the product and the containing portfolio are made available.

> At the moment, TagOptions can only be deactivated in the console.

Reviewer 2: Should we add this capability to the CDK L2? Seems very easy.

Author: We could revisit this, the issue is that you cannot create a tagOption in a disabled state, only on update can it be deactivated.  If we can have way to check for an on update in synth we could handle this.

Reviewer 2: We can just underscore that in the documentation I think.

### Notify on stack events

> Allows users to subscribe an AWS `SNS` topic to a provisiooned product's Cloudformation stack events.

Reviewer: ```suggestion
Allows users to subscribe an AWS `SNS` topic to a provisioned product's CloudFormation stack events.
```

> Allows users to subscribe an AWS `SNS` topic to a provisiooned product's Cloudformation stack events.

Author: ```suggestion
Allows users to subscribe an AWS `SNS` topic to a provisiooned product's CloudFormation stack events.
```

> When an end user provisions a product it creates a Cloudformation stack that notifies the subscribed topic on creation, edit, and delete events.

Reviewer: ```suggestion
When an end user provisions a product it creates a CloudFormation stack that notifies the subscribed topic on creation, edit, and delete events.
```

> When an end user provisions a product it creates a Cloudformation stack that notifies the subscribed topic on creation, edit, and delete events.

Author: ```suggestion
When an end user provisions a product it creates a CloudFormation stack that notifies the subscribed topic on creation, edit, and delete events.
```

An individual `SNS` topic may only have a single subscription to any given portfolio-product association.

```ts fixture=portfolio-product
import * as sns from '@aws-cdk/aws-sns';

const topic1 = new sns.Topic(this, 'Topic1');
portfolio.notifyOnStackEvents(product, topic1);

const topic2 = new sns.Topic(this, 'Topic2');
portfolio.notifyOnStackEvents(product, topic2, {
  description: 'description for topic2', // description is an optional field. 
});
```

### CloudFormation template parameters constraint

> CloudFormation template parameter constraints allow you to configure the provisioning parameters that are available to end users when they launch a product. Template constraint rules consist of one or more assertions that define the default and/or allowable values for a product’s provisioning parameters. You can configure multiple parameter constraints to govern the different provisioning parameters within your products. For example, a rule might define the `EC2` instance types that users can choose from when launching a product that includes one or more `EC2` instances. Parameter rules have an optional `condition` field that allow for rule application to consider conditional evaluations. If a `condition` is specified, all  assertions will be applied if the condition evaluates to true. For information on rule-specific intrinsic functions to define rule conditions and assertions, see [AWS Rule Functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-rules.html).

Reviewer 2: ```suggestion
CloudFormation template parameter constraints allow you to configure the provisioning parameters that are available to end users when they launch a product.
Template constraint rules consist of one or more assertions that define the default and/or allowable values for a product’s provisioning parameters.
You can configure multiple parameter constraints to govern the different provisioning parameters within your products.
For example, a rule might define the `EC2` instance types that users can choose from when launching a product that includes one or more `EC2` instances.
Parameter rules have an optional `condition` field that allow for rule application to consider conditional evaluations.
If a `condition` is specified, all  assertions will be applied if the condition evaluates to true.
For information on rule-specific intrinsic functions to define rule conditions and assertions,
see [AWS Rule Functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-rules.html).
```

### Set launch role

> Allows you to configure a specific `IAM` role that Service Catalog assumes on behalf of the end user when launching a product. By setting a launch role constraint, you can maintain least permissions for an end user when launching a product. For example, a launch role can grant permissions for specific resource creation like an `S3` bucket that the user. The launch role must be assumed by the Service Catalog principal. You can only have one launch role set for a portfolio-product association, and you cannot set a launch role on a product that already has a StackSets deployment configured.

Reviewer 2: ```suggestion
Allows you to configure a specific `IAM` role that Service Catalog assumes on behalf of the end user when launching a product.
By setting a launch role constraint, you can maintain least permissions for an end user when launching a product.
For example, a launch role can grant permissions for specific resource creation like an `S3` bucket that the user.
The launch role must be assumed by the Service Catalog principal.
You can only have one launch role set for a portfolio-product association,
and you cannot set a launch role on a product that already has a StackSets deployment configured.
```

```

See [Launch Constraint](https://docs.aws.amazon.com/servicecatalog/latest/adminguide/constraints-launch.html) documentation
to understand the permissions that launch roles need.

### Deploy with StackSets

A StackSets deployment constraint allows you to configure product deployment options using 
[AWS CloudFormation StackSets](https://docs.aws.amazon.com/servicecatalog/latest/adminguide/using-stacksets.html). 

> You can specify one or more accounts and regions into which stack instances will launch when the product is provisioned. There is an additional field `allowStackSetInstanceOperations` that sets ability for end users to create, edit, or delete the stacks created by the StackSet.

Reviewer 2: ```suggestion
You can specify one or more accounts and regions into which stack instances will launch when the product is provisioned.
There is an additional field `allowStackSetInstanceOperations` that sets ability for end users to create, edit, or delete the stacks created by the StackSet.
```

```

See [Launch Constraint](https://docs.aws.amazon.com/servicecatalog/latest/adminguide/constraints-launch.html) documentation
to understand the permissions that launch roles need.

### Deploy with StackSets

A StackSets deployment constraint allows you to configure product deployment options using 
[AWS CloudFormation StackSets](https://docs.aws.amazon.com/servicecatalog/latest/adminguide/using-stacksets.html). 

> You can specify one or more accounts and regions into which stack instances will launch when the product is provisioned. There is an additional 
> The field `allowStackSetInstanceOperations` configures ability for end users to create, edit, or delete the stacks created by the StackSet.

Author: ```suggestion
You can specify one or more accounts and regions into which stack instances will launch when the product is provisioned. There is an additional field `allowStackSetInstanceOperations` that sets ability for end users to create, edit, or delete the stacks created by the StackSet.
```
